### DISCUSSION TITLE
Thumb Combos produce "dash" instead of "undescore" in Terminal

### DISCUSSION THREAD

**Discussion by azzamsa** [manna-harbour/miryoku#223] [Q&A] [answered]
Hi.

I have used Miryoku for over a year. today I want to adopt `MIRYOKU_KLUDGE_THUMBCOMBOS=yes`. But unfortunately, hitting `RETURN +  TAB` produce "dash" instead of "underscore" in Terminal applications. I tested it on keyboardtester.com it produce `SHITF + -`. It works in every app but the terminal. I need to press `RETURN +  SHIFT + TAB` to get the "underscore".

I tried GNOME Terminal, Blackbox, Wezterm. all of them have the same behavior.

Any suggestion to solve this issue?

Thanks a lot  for Miryoku :heart:

**Comment by manna-harbour** [comment]
So, without `MIRYOKU_KLUDGE_THUMBCOMBOS=yes`, holding `Enter` for Sym and tapping `Tab` for underscore produces underscore in Gnome Terminal, but with `MIRYOKU_KLUDGE_THUMBCOMBOS=yes` the same keystrokes produce hyphen instead? How about when tapping other keys on Sym? Is this on Wayland?

**Reply by azzamsa** [reply]
Most of them work ` {&*(}:%^+~@#|(0-0-`, but the `0` and `-`. For both keys, I need to press the additional shift key.

Yes, I am on Fedora Silverblue 38 GNOME 44.2 Wayland.

**Comment by manna-harbour** [comment]
To summarise, you've observed that, in terminals on Wayland, shifted keycodes (e.g. underscore and close paren) work normally, except when combos are enabled for those keys.

Try merging with QMK master and also develop, to see if the issue has already been fixed upstream.

Wayland has a bug where consecutive keystrokes require a certain time interval in between to register correctly. See https://gitlab.gnome.org/GNOME/gnome-shell/-/issues/5890. However this isn't unique to Wayland, and is also an issue with many other systems.

QMK has `TAP_CODE_DELAY` for this purpose. See https://github.com/qmk/qmk_firmware/blob/master/docs/config_options.md#behaviors-that-can-be-configured. This is used when multiple consecutive key press and release events are generated by the firmware, e.g. with shifted keycodes.

Also, combos are handled before regular keycode processing.

I haven't had a chance to test or to examine the code, but it may be that when a combo is not fired, the queued events are handled in the combo code (and without use of `TAP_CODE_DELAY` for shifted keycodes) instead of being passed through for normal processing (which does use `TAP_CODE_DELAY`).

@sigprof and @sevanteri would have a better idea.

**Reply by sevanteri** [reply]
Seeing `ibus` mentioned in the gnome-shell issue reminds me that I also had issues with it and have disabled it from my machines. Can't remember specifically what was the issues but I'm pretty sure it was related to keycodes with modifiers.

What comes to QMK and combos, I know that if a key is part of a combo, there is a small delay (`COMBO_TERM`) before it is sent to the computer which might possibly cause issues.

> the queued events are handled in the combo code (and without use of TAP_CODE_DELAY for shifted keycodes) instead of being passed through for normal processing (which does use TAP_CODE_DELAY).

The queued `keyrecord_t`s are sent through the usual paths of `process_record` or `action_tapping_process` so if `TAP_CODE_DELAY` is in use, it should work normally. There is some extra handling for `TAP_CODE_DELAY` but I think that was to delay only once so that multiple delays wouldn't happen. Not sure how well that works nowadays, haven't used `TAP_CODE_DELAY` for a good while now.

**Reply by sevanteri** [reply]
Ah, I see the Miryoku has `COMBO_TERM` set to the same value as `TAPPING_TERM`. I would also try and set `COMBO_TERM` to some smaller value. I wouldn't be surprised if there's some timing issues when the values are the same and tapping keys are used in combos. Though is that even the case? And have there been any other issues with combos until now? I'm genuinely curious.

**Reply by manna-harbour** [reply]
> Ah, I see the Miryoku has COMBO_TERM set to the same value as TAPPING_TERM. I would also try and set COMBO_TERM to some smaller value. I wouldn't be surprised if there's some timing issues when the values are the same and tapping keys are used in combos. Though is that even the case? And have there been any other issues with combos until now? I'm genuinely curious.

These are thumb combos where 2 thumb keys are combo'd with a single thumb, so the long combo term is to make that more forgiving, and works because the keys are ordinarily hit individually with the same digit. I haven't seen any issues due to this.

In the case of the reported issue, a layer-tap is being held (also part of a non-triggered combo), and then the tapped key on that layer has a direct binding to a shifted keycode. But that key is also part of a thumb combo, although it's not being triggered in this case. The shift and the base keycode are being sent, but the shift isn't registering in the terminal. So behaviour seems to be as expected, other that the timing of the shift and base key making up the shifted keycode. And there's no issue when combos are disabled, so it seems to be an issue with processing of queued events of non-triggered combos.

**Reply by manna-harbour** [reply]
> The queued keyrecord_ts are sent through the usual paths of process_record or action_tapping_process so if TAP_CODE_DELAY is in use, it should work normally. There is some extra handling for TAP_CODE_DELAY but I think that was to delay only once so that multiple delays wouldn't happen. Not sure how well that works nowadays, haven't used TAP_CODE_DELAY for a good while now.

Thanks for the info!

@azzamsa That would be easy to confirm, just set `TAP_CODE_DELAY` to some very large value.

**Reply by azzamsa** [reply]
> @azzamsa That would be easy to confirm, just set TAP_CODE_DELAY to some very large value.

Any suggested value to start?

**Reply by manna-harbour** [reply]
The default is 100ms, so try 5 or 10 x that. Then you should be able to see the delay when tapping any symbols on Sym, and you'll be able to see if there is still a delay when building with combos activated.

**Reply by manna-harbour** [reply]
Oh, the docs say the default is 0! That's probably the issue then, although in that case it shouldn't make any difference if combos are enabled or not... Something like 10 or 20 should be enough in practice, but use something longer I'd you want to be able to see the difference.

**Reply by sigprof** [reply]
`TAP_CODE_DELAY` does not affect keycodes with modifiers like `S(KC_MINS)` — it is applied inside `tap_code()`, `tap_code16()` and in some other places where a release event is emitted immediately after a press event (e.g., when handling the tap case for `LT()` or `MT()`), but not between the report that changes the modifier state and the report that sends the basic keypress.  In other words, `TAP_CODE_DELAY` is the duration of a simulated tap (and in case of `send_string()` it is not even added between the release of one key and the press of the next key — that delay is controlled by the second argument of `send_string_with_delay()`).

If the keycode with modifiers is handled normally and reaches `process_action()`, the modifier change and the basic keypress should be sent in separate HID reports (the sequence of operations would be `add_weak_mods(mods); send_keyboard_report(); register_code(action.key.code)`).  The same thing would happen when using `register_code16()` or `tap_code16()` (in that case the `register_weak_mods()` helper would be used, which also calls `send_keyboard_report()` if any modifiers were specified).  So the delay between the modifier change and the keypress would be minimal (`USB_POLLING_INTERVAL_MS`, which defaults to 1 for some time).

(In fact there is some possibility of wrong event ordering if the modified keycode is not a plain key; e.g., in https://github.com/qmk/qmk_firmware/issues/21351 apparently that problem was observed with something like `G(KC_BTN1)`, and the same thing probably could happen with media keys; normal keys should not be affected like that though, because they would be sent over the same HID report as the modifier changes.)

**Reply by azzamsa** [reply]
I tried to put this code in my `custom_config.h`

```c
#undef TAP_CODE_DELAY
#define TAP_CODE_DELAY <number>
```

I can't find any ideal number, I tried 10, 20, 80, 160, 200. 
200 is too slow for the characters to appear.
20 seems okay, but still, I need to hold `ENTER` slightly longer before hitting 	` TAB` to get the underscore.
I don't type that slow, so it is a very unpleasant experience. Especially, for typing that can't be seen, such as password. I can't make sure it is a dash or underscore.  

For now, I think I will just disable the `MIRYOKU_KLUDGE_THUMBCOMBOS=yes`

**Reply by sevanteri** [reply, marked as answer]
Can you try and set `COMBO_TERM` to something lower than 200 (Without setting `TAP_CODE_DELAY` to anything)? The default is 50 which I think is still quite usable.

**Reply by azzamsa** [reply]
Isn't it already set to 200?

https://github.com/manna-harbour/miryoku_qmk/blob/e1307cc3a71e8f23c6f6c337764d6c854da009a6/users/manna-harbour_miryoku/config.h#L40

**Reply by sevanteri** [reply]
Yeah, I mean could you set it to something lower and see if it has any effects?

**Reply by azzamsa** [reply]
I tried 100. It feels better. the lower the number, the better I feel.
Now, I set it to 5. Huge improvement from before. But still, I need to slow down a tiny bit to get the underscore. 
I don't know if this has any impact to other key, I tested the combo and it still works fine.

```diff
modified   users/manna-harbour_miryoku/config.h
@@ -37,6 +37,6 @@
 // Thumb Combos
 #if defined (MIRYOKU_KLUDGE_THUMBCOMBOS)
   #define COMBO_COUNT 8
-  #define COMBO_TERM 200
+  #define COMBO_TERM 5
   #define EXTRA_SHORT_COMBOS
 #endif
modified   users/manna-harbour_miryoku/custom_config.h
```
